WIP: image stamping

vincepri opened the pull request:
This is an initial port from [wardroom](https://github.com/heptiolabs/wardroom) adding support for containerd in Ubuntu (RHEL to follow).

chuckha commented:
Thoughts on purging the wardroom name in favor of something more descriptive like 'kubernetes-image', 'kubeadm-image', or 'cluster-api-provider-aws-image'?

detiber commented on images/ansible/roles/kubernetes/defaults/main.yml:
```diff
@@ -0,0 +1,9 @@
+---
+kubernetes_version: "{{ kubernetes.version if kubernetes_version_defined else '1.9.3-00' }}"
```
Kill all the backwards compat for `kubernetes.*` variables here and switch to simple defaults.

Default for versions should probably be an empty string or undefined, and the role should default to latest available release version if not specified.

At some point we may want to consider a way to handle alpha/beta/rc releases as well, but that isn't a priority for the MVP.

detiber commented on images/ansible/roles/kubernetes/tasks/main.yml:
```diff
@@ -0,0 +1,23 @@
+  docker_image:
+    name: "{{ item }}"
+  with_items: kubernetes_cached_images
+  when: kubernetes_enable_cached_images | bool
```
Lets kill the cached image handling for now, especially since it is docker specific.

detiber commented on images/ansible/roles/kubernetes/templates/etc/systemd/system/kubelet.service:
```diff
@@ -0,0 +1,12 @@
+RestartSec=10
+
+[Install]
+WantedBy=multi-user.target
```
We shouldn't need this systemd unit file, it should be managed by the kubelet/kubeadm packages.

detiber commented on images/ansible/roles/kubernetes/templates/etc/systemd/system/kubelet.service.d/10-kubeadm.conf:
```diff
@@ -0,0 +1,7 @@
+Environment="KUBELET_NETWORK_ARGS=--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin"
+Environment="KUBELET_DNS_ARGS=--cluster-dns=10.96.0.10 --cluster-domain=cluster.local"
+ExecStart=
+ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_SYSTEM_PODS_ARGS $KUBELET_NETWORK_ARGS $KUBELET_DNS_ARGS $KUBELET_EXTRA_ARGS
```
For kubernetes v1.11+, we should be setting values in `/var/lib/kubelet/kubeadm-flags.env` insetad of overriding the unit files here.

detiber commented on images/ansible/roles/kubernetes/vars/main.yml:
```diff
@@ -0,0 +1,9 @@
+kubernetes_apt_key_url_defined: "{{ kubernetes is defined and 'apt_key_url' in kubernetes }}"
+kubernetes_apt_repo_string_defined: "{{ kubernetes is defined and 'apt_repo_string' in kubernetes }}"
+kubernetes_yum_baseurl_defined: "{{ kubernetes is defined and 'yum_baseurl' in kubernetes }}"
+kubernetes_yum_gpgkey_defined: "{{ kubernetes is defined and 'yum_gpgkey' in kubernetes }}"
```
This is all backward compat crud.

detiber commented:
@chuckha good point on the naming. @vincepri we'll also need to migrate over the python tool for sync'ing images. I'm also thinking that it might be a good idea to use that python tool for wrapping the interaction with packer for common use cases as well.

vincepri commented:
@detiber thanks for the review! fixed most comments and updated the PR. I didn't initially migrate the wardroom tool because we should be able to use packer directly to copy images to different regions [this line](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/25/files/2098247736d87f7ee3f19e906c3c5b3ef6a243b1#diff-117e24ee13e28cd08e452ab02d5d7c88R9) is a list of regions we want to copy the images to. Thoughts?

detiber commented:
> @detiber thanks for the review! fixed most comments and updated the PR. I didn't initially migrate the wardroom tool because we should be able to use packer directly to copy images to different regions this line is a list of regions we want to copy the images to. Thoughts?

Nice, I think that would work for the general user use case, but we will still probably want some tooling for the community AMI building. The wardroom tool for example will sync to all available regions in the client config and ensure the images are public.

vincepri commented on images/ansible/filter_plugins/kube.py:
```diff
@@ -0,0 +1,28 @@
+            else:
+                return version
+        if len(sub) == 2:
+            if platform.lower() == "redhat":
```
Just tested it, seems to be working with "redhat" as platform https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/25/commits/3856319315cc043929089ba85cd12b920593ee37

vincepri commented on images/ansible/roles/common/tasks/main.yml:
```diff
@@ -0,0 +1,6 @@
+  when: ansible_os_family == "Debian"
+
+- import_tasks: redhat.yml
+  when: ansible_os_family == "RedHat"
```
See https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/25/commits/3856319315cc043929089ba85cd12b920593ee37

detiber commented on images/ansible/roles/kubernetes/defaults/main.yml:
```diff
@@ -0,0 +1,3 @@
+---
+kubernetes_version: ""
+kubernetes_cni_version: ""
```
Does this behave correctly with the defaults here? It appears this may fail, since we are passing the value into the kube_platform_version filter.

detiber commented on images/packer/packer.json:
```diff
@@ -0,0 +1,103 @@
+  "variables": {
+    "aws_access_key": "",
+    "aws_secret_key": "",
+    "build_version": null,
```
I wonder if we should just nix build_version completely here and use `isotime` in it's place... @chuckha thoughts?

vincepri replied:
I'm +1 on this

detiber commented on images/packer/packer.json:
```diff
@@ -0,0 +1,103 @@
+    "aws_secret_key": "",
+    "build_version": null,
+    "kubernetes_version": "1.11.2-00",
+    "kubernetes_cni_version": "0.6.0-00",
```
Ideally, I'd like to see these variables default to "" and have the Ansible roles detect the latest available stable version, with these variables only being used for explicitly overriding the defaults. Other wise we're going to have to continually and explicitly rev this for every release.

vincepri replied:
If we opt for those value to be empty, should we change the `ami_name` and the tags so they don't include the Kubernetes version?

detiber replied:
doh, no, we still want to set those, otherwise we'll lose the ability to filter images by version.

vincepri replied:
Sounds good, I'll leave these in then and make the ansible default to `latest`.

detiber commented on images/README.md:
```diff
@@ -0,0 +1,110 @@
+
+| Variable | Default | Description |
+|----------|---------|-------------|
+| build_version | unset | A unique build version for the image |
```
Lets go ahead and kill build_version in favor of using a timestamp instead. That way we can avoid requiring a user to manually specify it for each image built.

We can always re-add it in the future if using a timestamp becomes an issue.

vincepri replied:
Sounds good

detiber commented on images/README.md:
```diff
@@ -0,0 +1,110 @@
+
+### Required Permissions to Build the AWS AMIs
+
+The [Packer documentation for the Amazon AMI builder](https://www.packer.io/docs/builders/amazon.html) supplies a suggested set of minimum permissions. However, Wardroom has been successfully tested with the following IAM permissions:
```
kill wardroom reference.

vincepri replied:
Done

detiber commented on images/README.md:
````diff
@@ -0,0 +1,110 @@
+        }
+    ]
+}
+```
````
Have you been able to test using these limited permissions? I'm wondering if the use of `ami_groups` or `ami_regions` adds additional requirements.

vincepri replied:
I've updated the permissions to match the official packer suggested ones. Which include permissions to modify images attributes, needed for ami_groups to function. The `ami_regions` should be good too.

detiber commented on images/packer/packer.json:
```diff
@@ -0,0 +1,103 @@
+      "ssh_username": "centos",
+      "tags": {
+        "build_version": "{{user `build_version`}}",
+        "source_ami": "{{user `centos_7_4_ami`}}",
```
I'm wondering if we can use `source_ami_filter` here instead to be able to automatically choose the newest base image rather than having to hard code it into a variable file like we are currently doing.

Thoughts?

vincepri replied:
Yeah. I was actually going to change it to use the official versions. The reason I left it untouched is that with the current behavior it's easier for folks to provide their own base image, which might contain different drivers/tweaks/security tools.

vincepri commented:
@detiber Fixed / replied to the other comments. I think this is ready to be merged in. 😊

chuckha commented:
Do these files need the licensing stanza at the top?

vincepri commented:
All done! Thanks @chuckha

detiber commented on images/ansible/filter_plugins/kube.py:
```diff
@@ -1,3 +1,17 @@
+# Copyright 2018 Heptio Inc.
```
s/Heptio/The Kubernetes Authors/

We should probably have this on all the Ansible yaml files as well.

vincepri commented:
@detiber License notices added, squashed commits.

detiber commented:
/ok-to-test
/lgtm

detiber commented:
/approve
